### ISSUE TITLE
Panic on Talos 1.6.7 caused by Cilium EBPF

### ISSUE THREAD

**Issue by bernardgut** [cilium/cilium#34700]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

equal or higher than v1.16.0 and lower than v1.17.0

### What happened?

Same config on VM works fine. 
On bare-metal : 

```
05/09/2024 13:12:30
================================================================================
05/09/2024 13:12:30
UBSAN: array-index-out-of-bounds in kernel/bpf/lpm_trie.c:194:14
05/09/2024 13:12:30
index 8 is out of range for type '__u8 [*]'
05/09/2024 13:12:30
CPU: 3 PID: 5346 Comm: cilium-agent Tainted: P O 6.6.43-talos #1
05/09/2024 13:12:30
Hardware name: Supermicro Super Server/H13SAE-MF, BIOS 2.0a 06/05/2024
05/09/2024 13:12:30
Call Trace:
05/09/2024 13:12:30
<TASK>
05/09/2024 13:12:30
dump_stack_lvl+0x47/0x60
05/09/2024 13:12:30
__ubsan_handle_out_of_bounds+0xc7/0x100
05/09/2024 13:12:30
[talos] controller failed {"component": "controller-runtime", "controller": "k8s.KubeletStaticPodController", "error": "error refreshing pod status: error fetching pod status: Get \"https://127.0.0.1:10250/pods/?timeout=30s\": remote error: tls: internal error"}
05/09/2024 13:12:31
longest_prefix_match.isra.0+0x18c/0x1d0
05/09/2024 13:12:31
trie_update_elem+0x152/0x300
05/09/2024 13:12:31
bpf_map_update_value+0xe2/0x220
05/09/2024 13:12:31
__sys_bpf+0x1afb/0x26f0
05/09/2024 13:12:31
__x64_sys_bpf+0x1e/0x30
05/09/2024 13:12:31
do_syscall_64+0x5a/0x80
05/09/2024 13:12:31
entry_SYSCALL_64_after_hwframe+0x78/0xe2
05/09/2024 13:12:31
RIP: 0033:0x40708e
05/09/2024 13:12:31
Code: 48 83 ec 38 e8 13 00 00 00 48 83 c4 38 5d c3 cc cc cc cc cc cc cc cc cc cc cc cc cc 49 89 f2 48 89 fa 48 89 ce 48 89 df 0f 05 <48> 3d 01 f0 ff ff 76 15 48 f7 d8 48 89 c1 48 c7 c0 ff ff ff ff 48
05/09/2024 13:12:31
RSP: 002b:000000c003604630 EFLAGS: 00000212 ORIG_RAX: 0000000000000141
05/09/2024 13:12:31
RAX: ffffffffffffffda RBX: 0000000000000002 RCX: 000000000040708e
05/09/2024 13:12:31
RDX: 0000000000000020 RSI: 000000c0036047b0 RDI: 0000000000000002
05/09/2024 13:12:31
RBP: 000000c003604670 R08: 0000000000000000 R09: 0000000000000000
05/09/2024 13:12:31
R10: 0000000000000000 R11: 0000000000000212 R12: 000000c003604360
05/09/2024 13:12:31
R13: 0000000000000000 R14: 000000c00237da40 R15: 00001fffffffffff
05/09/2024 13:12:31
</TASK>
05/09/2024 13:12:31
================================================================================
```

### How can we reproduce the issue?

1. Provision a single node bare-metal cluster on Talos 1.7.6 with the following Cilium config:

```
# disable the kube-proxy
kubeProxyReplacement: true
# stacks settings
routingMode: native
ipv4:
  enabled: true
ipv4NativeRoutingCIDR: 10.229.0.0/16 # podSubnet
enableIPv4BIGTCP: true
#enableIPv4Masquerade: false
autoDirectNodeRoutes: true
ipv6:
  enabled: true
enableIPv6BIGTCP: true
#enableIPv6Masquerade: false
ipv6NativeRoutingCIDR: fd00:0:a:1000::/64 # podSubnet
# optimizations
bpf:
  # TODO requires kernel 6.8+ Enable when available
  # datapathMode: netkit 
  masquerade: true
bandwidthManager:
  enabled: true
  bbr: true
# ipam:
#   mode: kubernetes
loadBalancer:
  algorithm: maglev
  mode: hybrid # https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#direct-server-return-dsr
  acceleration: best-effort # only enables XDP on devices that support it. Should work on HW devices. not on Wireguard
# if maglev is enabled:
maglev:
  tableSize: 4093 # default is 16381 but we have a small cluster. here we support up to 40 backends for 1 service
  hashSeed: 8vhAoJqAgVT0iVSP
# BGP
bgpControlPlane:
  enabled: true
# Ports Allowed
externalIPs:
  enabled: true
# Cilium
securityContext:
  capabilities:
    ciliumAgent:
      - CHOWN
      - KILL
      - NET_ADMIN
      - NET_RAW
      - IPC_LOCK
      - SYS_ADMIN
      - SYS_RESOURCE
      - DAC_OVERRIDE
      - FOWNER
      - SETUID
      - SETGID
    cleanCiliumState:
      - NET_ADMIN
      - SYS_ADMIN
      - SYS_RESOURCE
cgroup:
  hostRoot: /sys/fs/cgroup
  autoMount:
    enabled: false
# Enable Cilium Ingress Controller
ingressController:
  enabled: true
  default: true
  #loadbalancerMode: shared #TODO disable this if you go to IPv6
# Use KubePrism to access cluster API
k8sServiceHost: localhost
k8sServicePort: 7445
# OR
# Use IPv6 to access cluster API (for IPv6 only clusters)
#k8sServiceHost: ::1 # virtual IP
#k8sServicePort: 6443
# Enable Cilium Operator
operator:
  replicas: 1
# enable gatewayAPI
gatewayAPI:
  enabled: true
# Security
encryption:
  enabled: true
  type: wireguard

# Enable Hubble
...

```
2. watch as the node gets provisionned properly until Cilium Agent sends it into a boot loop

### Cilium Version

quay.io/cilium/cilium:v1.16.1@sha256:0b4a3ab41a4760d86b7fc945b8783747ba27f29dac30dd434d94f2c9e3679f39

### Kernel Version

6.6.43-talos

### Kubernetes Version

1.30.3

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bernardgut** [comment]
possibly related to https://github.com/cilium/cilium/issues/32812

**Comment by bernardgut** [comment]
After commenting these two lines :

```
loadbalancer:
  mode: hybrid # https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#direct-server-return-dsr
  acceleration: best-effort # only enables XDP on devices that support it. Should work on HW devices. not on Wireguard
```

The node doesnt go into a boot loop anymore. Still getting the following panic though:

```
05/09/2024 15:38:26
================================================================================
05/09/2024 15:38:26
UBSAN: array-index-out-of-bounds in kernel/bpf/lpm_trie.c:194:14
05/09/2024 15:38:26
index 8 is out of range for type '__u8 [*]'
05/09/2024 15:38:26
CPU: 17 PID: 5962 Comm: cilium-agent Tainted: P O 6.6.43-talos #1
05/09/2024 15:38:26
Hardware name: Supermicro Super Server/H13SAE-MF, BIOS 2.0a 06/05/2024
05/09/2024 15:38:26
Call Trace:
05/09/2024 15:38:26
<TASK>
05/09/2024 15:38:26
dump_stack_lvl+0x47/0x60
05/09/2024 15:38:26
__ubsan_handle_out_of_bounds+0xc7/0x100
05/09/2024 15:38:26
longest_prefix_match.isra.0+0x18c/0x1d0
05/09/2024 15:38:26
trie_update_elem+0x152/0x300
05/09/2024 15:38:26
bpf_map_update_value+0xe2/0x220
05/09/2024 15:38:26
__sys_bpf+0x1afb/0x26f0
05/09/2024 15:38:26
__x64_sys_bpf+0x1e/0x30
05/09/2024 15:38:26
do_syscall_64+0x5a/0x80
05/09/2024 15:38:26
entry_SYSCALL_64_after_hwframe+0x78/0xe2
05/09/2024 15:38:27
RIP: 0033:0x40708e
05/09/2024 15:38:27
Code: 48 83 ec 38 e8 13 00 00 00 48 83 c4 38 5d c3 cc cc cc cc cc cc cc cc cc cc cc cc cc 49 89 f2 48 89 fa 48 89 ce 48 89 df 0f 05 <48> 3d 01 f0 ff ff 76 15 48 f7 d8 48 89 c1 48 c7 c0 ff ff ff ff 48
05/09/2024 15:38:27
RSP: 002b:000000c003b8c198 EFLAGS: 00000206 ORIG_RAX: 0000000000000141
05/09/2024 15:38:27
RAX: ffffffffffffffda RBX: 0000000000000002 RCX: 000000000040708e
05/09/2024 15:38:27
RDX: 0000000000000020 RSI: 000000c003b8c318 RDI: 0000000000000002
05/09/2024 15:38:27
RBP: 000000c003b8c1d8 R08: 0000000000000000 R09: 0000000000000000
05/09/2024 15:38:27
R10: 0000000000000000 R11: 0000000000000206 R12: 000000c003b8bec8
05/09/2024 15:38:27
R13: 0000000000000000 R14: 000000c000d80fc0 R15: 0186000000006019
05/09/2024 15:38:27
</TASK>
05/09/2024 15:38:28
================================================================================
05/09/2024 15:38:36
eth0: renamed from tmpf2d63
```

cilium status reports everything fine but I havent tested yet.

**Comment by joestringer** [comment]
Hi @bernardgut , thanks for the report. This looks like an issue with the upstream Linux kernel. While Cilium can be useful to reproduce the issue, the underlying cause is not due to Cilium code. We can track the issue here in the Cilium project but ultimately this needs to be followed up with your upstream Linux maintainers and Talos distribution maintainers to pull in the fixed kernel version once the bug is addressed.

**Comment by samip5** [comment]
Technically, it is a Cilium related error as it's trying to use eBPF structs, that are not there anymore with the expected signature.

**Comment by joestringer** [comment]
I'm not disagreeing that Cilium can be used to reproduce the issue. However, the Linux kernel should not panic regardless of user input or the userspace programs run on top. If it does crash, that's a problem in the kernel and it needs to be fixed in the kernel.

**Comment by samip5** [comment]
> I'm not disagreeing that Cilium can be used to reproduce the issue. However, the Linux kernel should not panic regardless of user input or the userspace programs run on top. If it does crash, that's a problem in the kernel and it needs to be fixed in the kernel.

If the expectation of Cilium's usage of the bpf structs are not valid in all versions, I would be inclined to say problem is Cilium's usage instead but this has been fixed in 6.9 kernel.

Related: https://lore.kernel.org/all/202206281009.4332AA33@keescook/

**Comment by joestringer** [comment]
Since this has been fixed in the upstream Linux kernel, I'll close this issue. Feel free to continue to post any other details about this failure with relation to the kernel versions affected or kernel versions that fix this issue.
